### PR TITLE
docs for URI and CODE

### PR DESCRIPTION
**Pull request by drernie**
Documentation for
- "CODE" tab for Bucket and Packages views
- Quilt+ URI

~~Propose new '&catalog` fragment.~~

### CODE REVIEW

**Comment by drernie** [comment]
@fiskus @sir-sigurd Any thoughts? I'm sure @nl0 will want to weigh in, but I'd appreciate your feedback before then.

File: docs/Catalog/URI.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+In addition, it may contain the following optional components:
+
+- `<package>@<top_hash>`: The hash for this specific package, e.g.
+  `e21682f00929661879633a5128aaa27cc7bc1e2973d49d4c868a90f9fad9f34b`.
+- `<package>:tag`: The tag for this specific package. Currently, only the
+  `latest` tag is supported.  You may not specify both a top_hash and a tag.
+- `&path=<path>`: Fragment for the path to the object within the package, if
+  any, e.g.ß `CORD19.ipynb`.
+- `&catalog=<catalog>`: Fragment for the DNS name of catalog where this package
```
Comment (fiskus) [inline docs/Catalog/URI.md:56]: Although, I am not against adding the `catalog` parameter, I'm describing my doubts.
It looks like, package, top_hash and path describe WHERE data is located. And the data is expected to be there no matter what, until S3 is online, and the user pays for it. On the other hand, `catalog` describes HOW to get that data. And lifespan of the catalog is different from S3.
And it opens a question, should this type of information belong to URI.
* Basic HTTP auth, for example, is not a part of URI, it transmitted with HTTP Header
* But user/password for FTP is part of URI (ftp://user:pass@host)
* And, user/password for SSH too (ssh://user:pass@host)

So, considering this, I think, it's legit. But, maybe we can create separate scheme : for example, `quilt+catalog://{quilt+s3-uri}@catalog-name` → `quilt+catalog://quilt-example%23package%3Dakarve%2Fcord19%40e21682f00929661879633a5128aaa27cc7bc1e2973d49d4c868a90f9fad9f34b%26path%3DCORD19.ipynb%26catalog%3Dopen.quiltdata.com@catalog-name`
It seems like overengineering, though. And barely readable

**Comment by drernie** [comment]
Would it help to make the catalog part of the query string instead of the
fragment?

quilt+s3://bkt?catalog=dns.name#package=pre/suf

**Comment by fiskus** [comment]
> quilt+s3://bkt?catalog=dns.name#package=pre/suf

I don't have a rational explanation, but that looks more elegant for me. I don't know (or don't remember), what was the rationale behind using `#` hash instead of `?` query string in the first place.

File: docs/Catalog/URI.md (hunk 1 above)
Comment (drernie) [reply docs/Catalog/URI.md:56]: Moved catalog to seperate PR

**Comment by sir-sigurd** [comment]
> I don't know (or don't remember), what was the rationale behind using `#` hash instead of `?` query string in the first place.

AFAIR that was done because unlike fragment you need to encode querystring `/` in particular

File: docs/Catalog/URI.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+references.
+
+![Resolving URIs](../imgs/uri-resolve.png)
+
+### Syntax
+
+A Quilt+ URI most contain the following components:
+
+- `quilt+`: The scheme of the URI. This is always `quilt+`.
+- `s3://`: The protocol of the URI. This is currently `s3://`.
```
Comment (sir-sigurd) [inline docs/Catalog/URI.md:42]: well, AFAIK URI scheme is `quilt+s3`

File: docs/Catalog/URI.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+Catalog.  If your window is wide enough, there will be a "URI" button to the
+right of the search bar.  Clicking this button will display a dialog where you
+can paste a URI and "Resolve" it to navigate to the package or object it
+references.
+
+![Resolving URIs](../imgs/uri-resolve.png)
+
+### Syntax
+
+A Quilt+ URI most contain the following components:
```
Comment (sir-sigurd) [inline docs/Catalog/URI.md:39]:
```suggestion
A Quilt+ URI must contain the following components:
```

File: docs/Catalog/URI.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+- `<bucket>`: The name of the bucket containing the package or object, e.g.
+  `quilt-example`.
+- `#package=<package>`: A fragment for the name of the package containing the
+  object, e.g. `akarve/cord19`.
+
+In addition, it may contain the following optional components:
+
+- `<package>@<top_hash>`: The hash for this specific package, e.g.
+  `e21682f00929661879633a5128aaa27cc7bc1e2973d49d4c868a90f9fad9f34b`.
+- `<package>:tag`: The tag for this specific package. Currently, only the
```
Comment (sir-sigurd) [inline docs/Catalog/URI.md:52]:
```suggestion
- `<package>:<tag>`: The tag for this specific package. Currently, only the
```

File: docs/Catalog/URI.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  `latest` tag is supported.  You may not specify both a top_hash and a tag.
+- `&path=<path>`: Fragment for the path to the object within the package, if
+  any, e.g.ß `CORD19.ipynb`.
```
Comment (sir-sigurd) [inline docs/Catalog/URI.md:55]:
```suggestion
  any, e.g. `CORD19.ipynb`.
```

File: docs/Catalog/URI.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+In addition, Packages have a third tab that returns a Quilt+ URI:
+
+- URI (identifier) ![URI](../imgs/uri-uri.png)
+
+They all have a `copy` button that copies the code to the clipboard.
+
+## Quilt+ URIs
+
+Quilt+ URIs are a way to uniquely identify a package or object in the Quilt
```
Comment (nl0) [inline docs/Catalog/URI.md:17]: referencing packages vs objects seems misleading here, bc "objects" usually refer to "s3 objects", and quilt+ uris don't support those -- they support packages and "subpackages" (e.g. subsets of packages, down to individual package entries).

also, "Quilt bucket" seems somewhat vague as well -- these are "s3 buckets"

File: docs/Catalog/URI.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+catalog. They are used to reference packages and objects relative to a Quilt
+bucket.  For example:
```
Comment (nl0) [inline docs/Catalog/URI.md:19]:
```suggestion
bucket. For example:
```

File: docs/Catalog/URI.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+<!-- markdownlint-disable-next-line line-length -->
+`quilt+s3://quilt-example#package=akarve/cord19@e21682f00929661879633a5128aaa27cc7bc1e2973d49d4c868a90f9fad9f34b&path=CORD19.ipynb`
+
+The URI above references a specific version of the `CORD19.ipynb` notebook in
+the `akarve/cord19` package of the `quilt-example` bucket.
+
+### Catalog Usage
+
+URIs can be used to quickly navigate to a specific package or object from the
+Catalog.  If your window is wide enough, there will be a "URI" button to the
```
Comment (nl0) [inline docs/Catalog/URI.md:30]:
```suggestion
Catalog. If your window is wide enough, there will be a "URI" button to the
```

double spaces everywhere. isn't the linter supposed to catch these?

File: docs/Catalog/URI.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+### Syntax
+
+A Quilt+ URI most contain the following components:
+
+- `quilt+`: The scheme of the URI. This is always `quilt+`.
+- `s3://`: The protocol of the URI. This is currently `s3://`.
+- `<bucket>`: The name of the bucket containing the package or object, e.g.
+  `quilt-example`.
+- `#package=<package>`: A fragment for the name of the package containing the
+  object, e.g. `akarve/cord19`.
```
Comment (nl0) [inline docs/Catalog/URI.md:46]:
```suggestion
- `#package=<package>`: A fragment for the name of the package, e.g. `akarve/cord19`.
```

File: docs/Catalog/URI.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+In addition, it may contain the following optional components:
```
Comment (nl0) [inline docs/Catalog/URI.md:48]:
```suggestion
In addition, it _may_ contain the following optional components:
```

File: docs/Catalog/URI.md (hunk 3 above)
Comment (nl0) [inline docs/Catalog/URI.md:39]:
```suggestion
A Quilt+ URI _must_ contain the following components:
```

File: docs/Catalog/URI.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+- `<package>@<top_hash>`: The hash for this specific package, e.g.
```
Comment (nl0) [inline docs/Catalog/URI.md:50]: i'd group this under a single "specifier" term, which can be either a `@<hash>` or a `:<tag>` (optional, defaults to `:latest` when omitted)

File: docs/Catalog/URI.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+  `e21682f00929661879633a5128aaa27cc7bc1e2973d49d4c868a90f9fad9f34b`.
+- `<package>:tag`: The tag for this specific package. Currently, only the
+  `latest` tag is supported.  You may not specify both a top_hash and a tag.
+- `&path=<path>`: Fragment for the path to the object within the package, if
+  any, e.g.ß `CORD19.ipynb`.
+  was located.
```
Comment (nl0) [inline docs/Catalog/URI.md:56]: ? 🤔

File: docs/imgs/uri-resolve.png
Comment (nl0) [inline docs/imgs/uri-resolve.png:1]: maybe it's worth it to rename this to "Resolve a Quilt+ URI" in the catalog for consistency?

File: docs/Catalog/URI.md (hunk 2 above)
Comment (drernie) [reply docs/Catalog/URI.md:42]: At the moment, but we already have local and sharepoint repositories...

File: docs/Catalog/URI.md (hunk 3 above)
Comment (drernie) [reply docs/Catalog/URI.md:39]: Duplicate

File: docs/Catalog/URI.md (hunk 11 above)
Comment (drernie) [reply docs/Catalog/URI.md:49]: Make sense

File: docs/Catalog/URI.md (hunk 12 above)
Comment (drernie) [reply docs/Catalog/URI.md:55]: trimmed

File: docs/Catalog/URI.md (hunk 2 above)
Comment (drernie) [reply docs/Catalog/URI.md:41]: But yeah, I should use the term `scheme` correctly. Does this work?

File: docs/imgs/uri-resolve.png
Comment (drernie) [reply docs/imgs/uri-resolve.png:1]: Great idea!

File: docs/Catalog/URI.md (hunk 6 above)
Comment (drernie) [reply docs/Catalog/URI.md]: Switched to subpackage

File: docs/Catalog/URI.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+`quilt+s3://quilt-example#package=akarve/cord19@e21682f00929661879633a5128aaa27cc7bc1e2973d49d4c868a90f9fad9f34b&path=CORD19.ipynb`
+
+The URI above references a specific version of the `CORD19.ipynb` notebook in
+the `akarve/cord19` package of the `quilt-example` bucket.
+
+### Catalog Usage
+
+URIs can be used to quickly navigate to a specific package or object from the
+Catalog. If your window is wide enough, there will be a "URI" button to the
+right of the search bar.  Clicking this button will display a dialog where you
```
Comment (nl0) [inline docs/Catalog/URI.md:30]:
```suggestion
right of the search bar. Clicking this button will display a dialog where you
```

File: docs/Catalog/URI.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+![Resolving URIs](../imgs/uri-resolve.png)
+
+### Syntax
+
+A Quilt+ URI contains the following components:
+
+- `quilt+<protocol>`: The scheme of the URI. This always begins with `quilt+`.
+  Currently the only supported protocol is `s3`.
+- `<bucket>`: The name of the bucket containing the package or object, e.g.
```
Comment (nl0) [inline docs/Catalog/URI.md:42]:
```suggestion
- `<bucket>`: The name of the bucket containing the package, e.g.
```

File: docs/imgs/uri-resolve.png
Comment (nl0) [inline docs/imgs/uri-resolve.png:1]: since you've adjusted the caption, probably it's worth it to update the screenshot

**Review by nl0 — COMMENTED** [review]
looks good. some nitpicks inline

File: docs/imgs/uri-resolve.png
Comment (drernie) [reply docs/imgs/uri-resolve.png:1]: Done

**Review by nl0 — APPROVED** [review]

**Pull request merged**